aten/src/ATen/native/SoftMax.cpp

```diff
@@ -465,23 +479,62 @@ Tensor log_softmax(const Tensor& self, Dimname dim, optional<ScalarType> dtype)
   return at::log_softmax(self, dimname_to_position(self, dim), dtype);
 }
 
-Tensor masked_softmax_cpu(const Tensor& input, const Tensor& mask) {
+Tensor masked_softmax_cpu(const Tensor& input, int64_t dim, const Tensor& mask) {
   Tensor output = at::empty_like(input, input.options());
+  auto mask_ = mask.is_contiguous() ? mask : mask.contiguous();
   TORCH_CHECK(
       input.sizes() == mask.sizes(), "Mask shape should match input shape");
-  TORCH_CHECK(mask.is_contiguous(), "Mask should always be contiguous");
   TORCH_CHECK(
       mask.scalar_type() == ScalarType::Bool,
       "Mask should be a boolean tensor");
+
+  auto input_ = input.contiguous();
+  int64_t dim_ = maybe_wrap_dim(dim, input_.dim());
+
+  if (input_.dim() == 0) {
+    input_ = input_.view(1);
+  }
+
   AT_DISPATCH_FLOATING_TYPES_AND(
-      at::ScalarType::BFloat16, input.scalar_type(), "log_softmax", [&] {
+      at::ScalarType::BFloat16, input.scalar_type(), "masked_softmax", [&] {
         host_softmax<
             scalar_t,
             false /* LogSoftMax */,
             true /* MaskedSoftMax */>(
-            output, input, input.dim() - 1, mask.data_ptr<bool>());
+            output, input_, dim_, mask_.data_ptr<bool>());
       });
   return output;
 }
+
+Tensor masked_softmax_backward_cpu(
+    const Tensor& grad_,
+    const Tensor& output_,
+    int64_t dim_,
+    const Tensor& mask_) {
+  TORCH_CHECK(
+      grad_.sizes() == mask_.sizes(), "Mask shape should match grad shape");
+  TORCH_CHECK(
+      mask_.scalar_type() == ScalarType::Bool,
+      "Mask should be a boolean tensor");
+
+  int64_t dim = maybe_wrap_dim(dim_, grad_.dim());
+  auto grad = grad_.contiguous();
+  auto output = output_.contiguous();
+  auto mask = mask_.contiguous();
+
+  grad = grad.dim() == 0 ? grad.view(1) : grad;
+  output = output.dim() == 0 ? output.view(1) : output;
+  mask = mask.dim() == 0 ? mask.view(1) : mask;
+
+  Tensor grad_input = at::empty_like(grad, grad.options());
+  AT_DISPATCH_FLOATING_TYPES_AND(
+      at::ScalarType::BFloat16, grad.scalar_type(), "masked_softmax_backward", [&] {
+        host_softmax_backward<
+            scalar_t,
+            false /* LogSoftMax */,
+            true /* MaskedSoftmax */>(grad_input, grad, output, dim, mask.data_ptr<bool>());
+      });
+  return at::nan_to_num_(grad_input);
```

Comment:
Not entirely confident this is the right approach - I want to double check the logic here. Consider the following example from `test_masked_softmax_forward_with_nans`:

```
input
tensor([[-0.0883,  0.3420,  0.4112,  1.0051,  0.6003],
        [-0.8769,  0.9649, -0.1926,  0.1148, -0.1952],
        [-0.5215,  0.8718, -0.5610,  0.1900, -0.1660],
        [ 0.9107, -0.3660,  0.0853,  1.0006, -0.0962]], requires_grad=True)
mask
tensor([[False,  True, False,  True, False],
        [False,  True, False,  True, False],
        [False,  True, False,  True, False],
        [False,  True, False,  True, False]])
result
tensor([[   nan, 0.1978,    nan, 0.3510,    nan],
        [   nan, 0.3688,    nan, 0.1441,    nan],
        [   nan, 0.3360,    nan, 0.1554,    nan],
        [   nan, 0.0974,    nan, 0.3495,    nan]], grad_fn=<MaskedSoftmaxBackward0>)
expected
tensor([[   nan, 0.1978,    nan, 0.3510,    nan],
        [   nan, 0.3688,    nan, 0.1441,    nan],
        [   nan, 0.3360,    nan, 0.1554,    nan],
        [   nan, 0.0974,    nan, 0.3495,    nan]], grad_fn=<SoftmaxBackward0>)
```

Note two things:

1) Getting these `nan`s in  requires dividing by zero to achieve (e.g. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SoftMax.cpp#L200). The test works fine locally, but then in the test suite, it complains of dividing by zero 

2) The `at::nan_to_num` is necessary in order to match the gradient (otherwise, the 0's are replaced by `nan`s:
```
input_ref.grad
tensor([[ 0.0000, -0.1066,  0.0000, -0.0283,  0.0000],
        [ 0.0000,  0.2753,  0.0000, -0.0099,  0.0000],
        [ 0.0000, -0.0511,  0.0000,  0.1728,  0.0000],
        [ 0.0000, -0.1175,  0.0000, -0.1346,  0.0000]])
```
I checked the logic and even compared all the values being set in the backwards gradient from `Softmax.cpp` (where I also applied the same `nan_to_num` function), and they were identical. I'm wondering if the structured kernel takes care of this somehow by automatically converting nan's in the backwards gradient to 0?